TS3Bot.py

```diff
@@ -55,168 +43,16 @@ def __init__(self, db_name):
-                except Exception as ex:
-                    exres = exception_handler(ex)
-        if reinit:
-            self.init()
-        return res, exres
-
-    def close(self):
-        self.ts3exec(lambda tc: tc.close())
-
-    def copy(self):
-        tsc = ThreadsafeTSConnection(self._user, self._password, self._host, self._port, self._keepalive_interval, self._server_id, None)
-        # make sure to
-        # 1. not pass bot_nickname to the constructor, or the child (copy) would call forceRename and attempt to kick the parent
-        # 2. gently rename the copy afterwards
-        tsc.gentleRename(self._bot_nickname)
-        return tsc
-
-    def gentleRename(self, nickname):
-        """
-        Renames self to nickname, but attaches a running counter
-        to the name if the nickname is already taken.
-        """
-        i = 1
-        new_nick = "%s(%d)" % (nickname, i)
-        while not self.ts3exec(lambda tc: tc.query("clientfind", pattern=new_nick).first(), signal_exception_handler)[1]:
-            i += 1
-            new_nick = "%s(%d)" % (nickname, i)
-        new_nick = "%s(%d)" % (nickname, i)
-        self.ts3exec(lambda tc: tc.exec_("clientupdate", client_nickname=new_nick))
-        self._bot_nickname = new_nick
-        return self._bot_nickname
-
-    def forceRename(self, nickname):
-        """
-        Attempts to forcefully rename self.
-        If the chosen nickname is already taken, the bot will attempt to kick that user.
-        If that fails the bot will fall back to gentle renaming itself.
-        """
-        imposter, free = self.ts3exec(lambda tc: tc.query("clientfind", pattern=nickname).first(), signal_exception_handler)  # check if nickname is already in use
-        if not free:  # error occurs if no such user was found -> catching no exception means the name is taken
-            _, ex = self.ts3exec(lambda tc: tc.exec_("clientkick", reasonid=5, reasonmsg="Reserved Nickname", clid=imposter.get("clid")), signal_exception_handler)
-            if ex:
-                log.warning(
-                    "Renaming self to '%s' after kicking existing user with reserved name failed."
-                    " Warning: this usually only happens for serverquery logins, meaning you are running multiple bots or you"
-                    " are having stale logins from crashed bot instances on your server. Only restarts can solve the latter.",
-                    nickname)
-            else:
-                log.info("Kicked user who was using the reserved registration bot name '%s'.", nickname)
-            nickname = self.gentleRename(nickname)
-            log.info("Renamed self to '%s'.", nickname)
-        else:
-            self.ts3exec(lambda tc: tc.exec_("clientupdate", client_nickname=nickname))
-            log.info("Forcefully renamed self to '%s'.", nickname)
-        self._bot_nickname = nickname
-        return self._bot_nickname
-
-
 class Bot:
+    _ts_connection: ThreadsafeTSConnection
```

Comment:
this creates a static variable. Python implicitly declares attributes by assigning them from anywhere, preferably the constructor.

Reply:
I was trying to add type hinting, how whould you do it ?

Reply:
fixed.